blog/tidy-collaborative-filtering/index.qmd

```diff
@@ -0,0 +1,214 @@
+
+The main advantage of content-based filtering is that it can make recommendations for new users without any interaction history or recommend new items to users. The disadvantages include: (i) training needs a lot of users and item examples for reliable results; (ii) tuning might be much harder in practice than collaborative filtering; (iii) missing information might be a problem since there is no clear solution how to treat missingness in user or item characteristics.
+
+## Hybrid Recommenders
+
+Hybrid recommender systems combine both collaborative and content-based filtering to overcome the challenges of each individual approach. There are different hybridization techniques available, e.g., combining the scores of different components (weighted), choosing among different compontent (switching), following strict priority rules (cascading), presenting outputs from different components at the same time (mixed), etc.
+
+# Train Collaborative Filtering Recommenders in R
+
+For this post, we rely on the `tidyverse` [@Wickham2019] family of packages, `scales` [@scales] for scale functions for visualization, and `recommenderlab` [@recommenderlab] - a package that provides an infrastructure to develop and test collaborative filerting recommender algorithms.  
+```{r, message = FALSE, warning = FALSE}
+library(tidyverse)
+library(scales)
+library(recommenderlab)
+```
+
+```{r, include = FALSE}
+wikifolio_portfolios <- read_rds("../data/wikifolio_portfolios.rds")
+```
+
+I load a dataset with stock holdings of investable wikifolios at the beginning of 2023. Reach out if you want to get access to the data. 
+
+```{r}
+glimpse(wikifolio_portfolios)
+```
+
+First, I convert the long data to a binary rating matrix where only non-NA values are stored explicitly and NA values are represented by dots for efficiency. 
+
+```{r}
+binary_rating_matrix <- wikifolio_portfolios |>
+  pivot_wider(id_cols = wikifolio,
+              names_from = stock,
+              values_from = in_portfolio,
+              values_fill = list(in_portfolio = 0)) |>
+  select(-wikifolio) |>
+  as.matrix() |>
+  as("binaryRatingMatrix")
+binary_rating_matrix
+```
+
+As in our book chapter on factor selection via machine learning, I perform cross-validation and split the data into training and test data. The training sample constitute 80% of the data and I perform $k$-times recursive estimation of the cross validation with $k=5$. Note that the parameter `given=-1` means that all but one randomly selected item is ignored for the evaluation.
```

Comment:
- [factor selection via machine learning](https://www.tidy-finance.org/factor-selection-via-machine-learning.html)
- "perform recursive estimation of the cross-validation" sounds a bit odd. "I perform 5-fold cross-validation" would be the common way to express this. 
- You probably want to highlight what the hyperparameters are that are trained, right now it is a bit unclear to me what it actually is that is chosen.

Reply:
I improved the description of the testing procedure. I don't think that this is the right location to talk about hyperparameters because they depend on the algorithm (if there are any).